keyedcache/api.go

```diff
@@ -0,0 +1,153 @@
+// Package keyedcache provides a thread-safe cache to hold clean or dirty
+// objects identified by opaque keys in a bounded amount of memory.
+
+package keyedcache
+
+import (
+	"sync"
+)
+
+// Just ignore this for the time being, other then noting that a pointer to one
+// of these gets passed through requests into the keyed cache and down into the
+// callbacks.
+//
+type ReqContext interface {
+}
+
+// Objects in the cache are identified by a key, which is unique for each
+// object.  The key must support the following operations:
+//
+//   o Equal() -- returns true if the two keys refer to the same object
+//   o Hash() -- return a hash value based on the key (two keys that are equal
+//     must have the same hash value)
+//   o SizeOf() -- return memory used by the key (to track memory consumption);
+//     all keys are assumed to be the same size.
+//
+type Key interface {
+	Equal(key *Key) bool
+	Hash() uint64
+	SizeOf() uint64
+}
+
+// Objects in the cache are "allocated" by calling a NewObject function supplied
+// by the client when it created the cache.  New objects do not have an
+// indentity until they are assigned one by a lookup in the cache.
+//
+// Objects must support the following operations.  These methods are used
+// exclusively by the keyedcache; clients who get an object from the cache
+// should not call them, instead they should call methods provided by the cache
+// to flush or invalidate objects:
+//
+//   o Read() -- called to initialize an object when its assigned an identity;
```

Comment:
Is this a "Load()" method? Wouldn't we need to have a callback mechanism to enable the generic cache to be able to perform the type-specific "load" operation?